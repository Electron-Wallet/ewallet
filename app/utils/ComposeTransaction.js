/* @flow */
'use strict';

// import AbstractMethod from './AbstractMethod';
import Discovery from 'utils/helpers/Discovery';
// import * as UI from '../../constants/ui';
import { getCoinInfoByCurrency } from 'utils/data/CoinInfo';
import { validateParams } from 'utils/helpers/paramsValidator';
import { resolveAfter } from 'utils/promiseUtils';
import { formatAmount } from 'utils/formatUtils';
import { NO_COIN_INFO } from 'utils/constants/errors';

import BlockBook, { create as createBackend } from 'utils/backend';
import Account from 'utils/account';
import TransactionComposer from 'utils/tx/TransactionComposer';
import {
    validateHDOutput,
    inputToTrezor,
    outputToTrezor,
    getReferencedTransactions,
    transformReferencedTransactions,
} from 'utils/tx';
import * as helper from 'utils/helpers/signtx';

// import { UiMessage } from '../../message/builder';

import type { CoinInfo, UiPromiseResponse } from 'flowtype';
import type { CoreMessage } from 'utils/types';
import type { SignedTx } from 'utils/types/trezor';

import type {
    BuildTxOutputRequest,
    BuildTxResult,
} from 'hd-wallet';
import type { Options } from './backend/BlockBook';

type Params = {
    outputs: Array<BuildTxOutputRequest>,
    coinInfo: CoinInfo,
    push: boolean,
}

export default class ComposeTransaction  {
    params: Params;
    backend: BlockBook;
    discovery: ?Discovery;
    composer: TransactionComposer;

    constructor(options: Options) {
        // super(message);
        // this.requiredPermissions = ['read', 'write'];

        // validate incoming parameters
        validateParams(options, [
            { name: 'outputs', type: 'array', obligatory: true },
            { name: 'coin', type: 'string', obligatory: true },
            { name: 'push', type: 'boolean' },
        ]);

        const coinInfo: ?CoinInfo = getCoinInfoByCurrency(options.coin);
        if (!coinInfo) {
            throw NO_COIN_INFO;
        }

        // set required firmware from coinInfo support
        // this.requiredFirmware = [ coinInfo.support.trezor1, coinInfo.support.trezor2 ];

        // validate each output and transform into hd-wallet format
        const outputs: Array<BuildTxOutputRequest> = [];
        let total: number = 0;
        options.outputs.forEach(out => {
            const output = validateHDOutput(out, coinInfo);
            if (typeof output.amount === 'number') {
                total += output.amount;
            }
            outputs.push(output);
        });

        const sendMax: boolean = outputs.find(o => o.type === 'send-max') !== undefined;

        // there should be only one output when using send-max option
        if (sendMax && outputs.length > 1) {
            throw new Error('Only one output allowed when using "send-max" option.');
        }

        // if outputs contains regular items
        // check if total amount is not lower than dust limit
        if (outputs.find(o => o.type === 'complete') !== undefined && total <= coinInfo.dustLimit) {
            throw new Error('Total amount is too low.');
        }

        if (sendMax) {
            this.info = 'Send maximum amount';
        } else {
            this.info = `Send ${ formatAmount(total, coinInfo) }`;
        }

        this.params = {
            outputs,
            coinInfo,
            push: options.hasOwnProperty('push') ? options.push : false,
        };
    }

    async run(): Promise<SignedTx> {
        // initialize backend
        this.backend = await createBackend(this.params.coinInfo);

        // discover accounts and run
        const account = await this._getAccount();
        if (account instanceof Account) {
            // wait for fee selection
            const response: string | SignedTx = await this._getFee(account);
            if (typeof response === 'string') {
                // back to account selection
                return await this.run();
            } else {
                return response;
            }
        } else {
            throw new Error(account.error);
        }
    }

    async _getAccount(): Promise<Account | { error: string }> {
        // 替换自己的实现
        const discovery: Discovery = this.discovery || new Discovery({
            getHDNode: this.device.getCommands().getHDNode.bind(this.device.getCommands()),
            coinInfo: this.params.coinInfo,
            backend: this.backend,
        });

      const initPromise: Deferred<void> = createDeferred();

        discovery.on('update', (accounts: Array<Account>) => {
          console.log('discovery.on update')
        });

        discovery.on('complete', (accounts: Array<Account>) => {
          console.log('discovery.on complete');
          initPromise.resolve();
        });

        if (!this.discovery) {
            this.discovery = discovery;
        }
        discovery.start();

        // 增加 permise， 在 utils event complate 中完成
        await initPromise.promise;
        discovery.removeAllListeners();
        discovery.stop();

        if (discovery.accounts.length === 0) {
          return 'not found account'
        }

        return discovery.accounts[0];
    }

    async _getFee(account: Account): Promise<string | SignedTx> {
        if (this.composer) { this.composer.dispose(); }

        const composer: TransactionComposer = new TransactionComposer(account, this.params.outputs);
        await composer.init(this.backend);
        this.composer = composer;

        const hasFunds: boolean = await composer.composeAllFeeLevels();
        if (!hasFunds) {
            return 'Not has Funds';
        }

        // 外部传入
        const feeLevel = this.composer.feeLevels[0];
        return await this._send(feeLevel.name);
    }

    async _send(feeLevel: string): Promise<SignedTx> {
        const tx: BuildTxResult = this.composer.composed[feeLevel];

        if (tx.type !== 'final') throw new Error('Trying to sign unfinished tx');

        const bjsRefTxs = await this.backend.loadTransactions(getReferencedTransactions(tx.transaction.inputs));
        const refTxs = transformReferencedTransactions(bjsRefTxs);

        const coinInfo: CoinInfo = this.composer.account.coinInfo;

        const response = await helper.signTx(
            this.device.getCommands().typedCall.bind(this.device.getCommands()),
            tx.transaction.inputs.map(inp => inputToTrezor(inp, 0)),
            tx.transaction.outputs.sorted.map(out => outputToTrezor(out, coinInfo)),
            refTxs,
            coinInfo,
        );

        if (this.params.push) {
            const txid: string = await this.backend.sendTransactionHex(response.serializedTx);
            return {
                ...response,
                txid,
            };
        }

        return response;
    }

    dispose() {
        if (this.discovery) {
            const d = this.discovery;
            d.stop();
            d.removeAllListeners();
        }

        if (this.composer) {
            this.composer.dispose();
        }
    }
}
