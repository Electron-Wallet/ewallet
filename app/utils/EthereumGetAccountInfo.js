/* @flow */
'use strict';

import { validateParams } from './helpers/paramsValidator';
import { getEthereumNetwork } from 'utils/data/CoinInfo';
import Discovery from './helpers/Discovery';
import { NO_COIN_INFO } from 'constants/errors';

import BlockBook, { create as createBackend } from 'utils/backend';
import type { CoreMessage, EthereumNetworkInfo } from 'utils/types';
import type { EthereumAccount } from 'utils/types/account';
import type { Session as TrezorSession } from 'trezor.js';
import type { GetAccountInfoOptions } from './GetAccountInfo';
import Account from './account';

type Params = {
  account: EthereumAccount,
  coinInfo: EthereumNetworkInfo,
  // bundledResponse: boolean,
}

export type EthereumGetAccountInfoOptions = {
  coin?: string,
  account: EthereumAccount,
  session: TrezorSession,
};

export default class EthereumGetAccountInfo {
  params: Params;
  confirmed: boolean = false;
  backend: BlockBook;
  discovery: ?Discovery;

  constructor(options: EthereumGetAccountInfoOptions) {
    this.info = 'Export ethereum account info';
    this.useDevice = false;

    // const payload: Object = message.payload;
    // let bundledResponse: boolean = true;
    // // create a bundle with only one batch
    // if (!options.hasOwnProperty('accounts')) {
    //   options.accounts = [...options.account];
    //   bundledResponse = false;
    // }

    // validate incoming parameters
    validateParams(options, [
      { name: 'account', type: 'object', obligatory: true },
      { name: 'coin', type: 'string', obligatory: true }
    ]);

    validateParams(options.account, [
      { name: 'address', type: 'string', obligatory: true },
      { name: 'block', type: 'number', obligatory: true },
      { name: 'transactions', type: 'number', obligatory: true }
    ]);

    const network: ?EthereumNetworkInfo = getEthereumNetwork(options.coin);
    if (!network) {
      throw NO_COIN_INFO;
    }

    this.params = {
      account: options.account,
      coinInfo: network,
      // bundledResponse
    };
  }

  async run(): Promise<EthereumAccount | Array<EthereumAccount>> {
    // initialize backend
    this.backend = await createBackend(this.params.coinInfo);

    const blockchain = this.backend.blockchain;
    const { height } = await blockchain.lookupSyncStatus();

    const account = this.params.account;
    const method = 'getAddressHistory';
    const params = [
      [account.address],
      {
        start: height,
        end: account.block,
        from: 0,
        to: 0,
        queryMempol: false
      }
    ];
    const socket = await blockchain.socket.promise;
    const confirmed = await socket.send({ method, params });

    return {
      address: account.address,
      transactions: confirmed.totalCount,
      block: height,
      balance: '0', // TODO: fetch balance from blockbook
      availableBalance: '0', // TODO: fetch balance from blockbook
      nonce: 0 // TODO: fetch nonce from blockbook
    };

    /*
    // This will be useful for BTC-like accounts (multi addresses)
    const addresses: Array<string> = this.params.accounts.map(a => a.address);

    const socket = await blockchain.socket.promise;
    const method = 'getAddressHistory';
    const params = [
        addresses,
        {
            start: height,
            end: 0,
            from: 0,
            to: 0,
            queryMempol: true
        }
    ];

    const response = await socket.send({method, params});
    return {
        address: addresses[0],
        transactions: response.totalCount,
        block: height
    }
    */
  }
}
