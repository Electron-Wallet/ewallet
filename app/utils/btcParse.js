/* @flow */

const _ = require('lodash');

const Utils = () => {};

Utils.COIN = {
  bch: {
    name: 'bch',
    toSatoshis: 100000000,
    maxDecimals: 8,
    minDecimals: 8,
  },
  btc: {
    name: 'btc',
    toSatoshis: 100000000,
    maxDecimals: 8,
    minDecimals: 8,
  },
  bit: {
    name: 'bit',
    toSatoshis: 100,
    maxDecimals: 2,
    minDecimals: 2,
  },
}

Utils.UNITS2 = {
  btc: 100000000,
  bit: 100,
  sat: 1,
}

export const renderAmount = (satoshis, coin, opts) => {
  function clipDecimals(number, decimals) {
    const x = number.toString().split('.');
    const d = (x[1] || '0').substring(0, decimals);
    return parseFloat(`${x[0]}.${d}`)
  }

  function addSeparators(nStr, thousands, decimal, minDecimals) {
    nStr = nStr.replace('.', decimal) // eslint-disable-line
    const x = nStr.split(decimal);
    let x0 = x[0];
    let x1 = x[1];

    x1 = _.dropRightWhile(x1, (n, i) => {
      return n === '0' && i >= minDecimals;
    }).join('');
    const x2 = x.length > 1 ? decimal + x1 : '';

    x0 = x0.replace(/\B(?=(\d{3})+(?!\d))/g, thousands);
    return x0 + x2;
  }

  opts = opts || {}; // eslint-disable-line

  coin = coin || 'btc';  // eslint-disable-line
  const u = Utils.COIN[coin] || Utils.COIN.btc;
  const amount = clipDecimals((satoshis / u.toSatoshis), u.maxDecimals).toFixed(u.maxDecimals)
  return `${addSeparators(
    amount,
    opts.thousandsSeparator || ',',
    opts.decimalSeparator || '.',
    u.minDecimals,
  )} ${u.name}`
};

export const parseAmount = (text) => {
  if (!_.isString(text)) {
    text = text.toString() // eslint-disable-line
  }

  const regex = `^(\\d*(\\.\\d{0,8})?)\\s*(${_.keys(Utils.UNITS2).join('|')})?$`
  const match = new RegExp(regex, 'i').exec(text.trim())

  if (!match || match.length === 0) throw new Error('Invalid amount')

  const amount = parseFloat(match[1])
  if (!_.isNumber(amount) || _.isNaN(amount)) throw new Error('Invalid amount')

  const unit = (match[3] || 'sat').toLowerCase()
  const rate = Utils.UNITS2[unit]
  if (!rate) throw new Error('Invalid unit')

  const amountSat = parseFloat((amount * rate).toPrecision(12))
  if (amountSat !== Math.round(amountSat)) throw new Error('Invalid amount')

  return amountSat
};

