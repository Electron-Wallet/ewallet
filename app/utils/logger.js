/* @flow */
import Logger from "js-logger";

Logger.useDefaults();

if (process.env.NODE_ENV === 'development') {
  Logger.setLevel(Logger.DEBUG);
} else if (process.env.DEBUG_PROD === 'true') {
  Logger.setLevel(Logger.INFO);
} else {
  Logger.setLevel(Logger.WARN);
}

export default Logger;
