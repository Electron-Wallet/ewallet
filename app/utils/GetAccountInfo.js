/* @flow */
'use strict';

// import AbstractMethod from './AbstractMethod';
import { validateParams, validateCoinPath } from './helpers/paramsValidator';
import Discovery from 'utils/helpers/Discovery';
// import * as UI from '../../constants/ui';
import { NO_COIN_INFO } from 'constants/errors';
import * as hdnodeUtils from 'utils/hdnode';
import {
    isMultisigPath,
    isSegwitPath,
    isBech32Path,
    validatePath,
    getAccountLabel,
    getSerializedPath,
    getScriptType,
} from 'utils/pathUtils';
import { create as createDeferred } from 'utils/deferred';

import Account, { create as createAccount } from 'utils/account';
import BlockBook, { create as createBackend } from 'utils/backend';
import { getCoinInfoByCurrency, fixCoinInfoNetwork, getCoinInfoFromPath } from 'utils/data/CoinInfo';
// import { UiMessage } from '../../message/builder';
import type {
  CoinInfo,
  // UiPromiseResponse,
} from 'flowtype';
import type { AccountInfo, HDNodeResponse } from 'utils/types/trezor';
import type { Deferred, CoreMessage } from 'utils/types';
import type { Options } from './backend/BlockBook';

import { getSegwitNetwork, getBech32Network } from 'utils/data/CoinInfo';

type Params = {
    path: ?Array<number>,
    xpub: ?string,
    coinInfo: CoinInfo,
}

type Response = AccountInfo | {
    error: string,
}

export default class GetAccountInfo {
    params: Params;
    confirmed: boolean = false;
    backend: BlockBook;
    discovery: ?Discovery;

    constructor(options: Options) {

        this.session = options.session;

        // super(message);
        this.requiredPermissions = ['read'];
        this.info = 'Export account info';

        // const payload: Object = message.payload;

        // validate incoming parameters
        validateParams(options, [
            { name: 'coin', type: 'string' },
            { name: 'xpub', type: 'string' },
            { name: 'crossChain', type: 'boolean' },
        ]);

        let path: Array<number>;
        let coinInfo: ?CoinInfo;
        if (options.coin) {
            coinInfo = getCoinInfoByCurrency(options.coin);
        }

        if (options.path) {
            path = validatePath(options.path, 3, true);
            if (!coinInfo) {
                coinInfo = getCoinInfoFromPath(path);
            } else if (!options.crossChain) {
                validateCoinPath(coinInfo, path);
            }
        }

        // if there is no coinInfo at this point return error
        if (!coinInfo) {
            throw NO_COIN_INFO;
        } else {
            // check required firmware with coinInfo support
            this.requiredFirmware = [ coinInfo.support.trezor1, coinInfo.support.trezor2 ];
        }

        // delete payload.path;
        // payload.xpub = 'ypub6XKbB5DSkq8Royg8isNtGktj6bmEfGJXDs83Ad5CZ5tpDV8QofwSWQFTWP2Pv24vNdrPhquehL7vRMvSTj2GpKv6UaTQCBKZALm6RJAmxG6'
        // payload.xpub = 'xpub6BiVtCpG9fQQNBuKZoKzhzmENDKdCeXQsNVPF2Ynt8rhyYznmPURQNDmnNnX9SYahZ1DVTaNtsh3pJ4b2jKvsZhpv2oVj76YETCGztKJ3LM'

        this.params = {
            path: path,
            xpub: options.xpub,
            coinInfo,
        };

        this._getHDNode = this._getHDNode.bind(this);
        this.getBitcoinHDNode = this.getBitcoinHDNode.bind(this);
    }

    async confirmation(): Promise<boolean> {
        if (this.confirmed) return true;
        // wait for popup window
        // await this.getPopupPromise().promise;
        // initialize user response promise
        // const uiPromise = this.createUiPromise(UI.RECEIVE_CONFIRMATION, this.device);

        let label: string;
        if (this.params.path) {
            label = getAccountLabel(this.params.path, this.params.coinInfo);
        } else if (this.params.xpub) {
            label = `Export ${ this.params.coinInfo.label } account for public key <span>${ this.params.xpub }</span>`;
        } else {
            return true;
        }

        // request confirmation view
        // this.postMessage(new UiMessage(UI.REQUEST_CONFIRMATION, {
        //     view: 'export-account-info',
        //     label,
        // }));

        // wait for user action
        // const uiResp: UiPromiseResponse = await uiPromise.promise;
        // const resp: string = uiResp.payload;

        // this.confirmed = (resp === 'true');
        this.confirmed = true;
        return this.confirmed;
    }

    async run(): Promise<Response> {
        // initialize backend
        this.backend = await createBackend(this.params.coinInfo);

        if (this.params.path) {
            return await this._getAccountFromPath(this.params.path);
        } else if (this.params.xpub) {
            return await this._getAccountFromPublicKey();
        } else {
            return await this._getAccountFromDiscovery();
        }
    }

    async _getAccountFromPath(path: Array<number>): Promise<Response> {
        const coinInfo: CoinInfo = fixCoinInfoNetwork(this.params.coinInfo, path);
        const node: HDNodeResponse = await this._getHDNode(path, coinInfo);
        const account = createAccount(path, node.xpub, coinInfo);

        const discovery: Discovery = this.discovery = new Discovery({
            // getHDNode: this.device.getCommands().getHDNode.bind(this.device.getCommands()),
            getHDNode: this._getHDNode,
            coinInfo: this.params.coinInfo,
            backend: this.backend,
            loadInfo: false,
        });

        await discovery.getAccountInfo(account);
        return this._response(account);
    }

    async _getAccountFromPublicKey(): Promise<Response> {
        // const discovery: Discovery = this.discovery = new Discovery({
        //     getHDNode: this.device.getCommands().getHDNode.bind(this.device.getCommands()),
        //     coinInfo: this.params.coinInfo,
        //     backend: this.backend,
        //     loadInfo: false,
        // });
        //
        // const deferred: Deferred<Response> = createDeferred('account_discovery');
        // discovery.on('update', async (accounts: Array<Account>) => {
        //     const account = accounts.find(a => a.xpub === this.params.xpub);
        //     if (account) {
        //         discovery.removeAllListeners();
        //         discovery.completed = true;
        //
        //         await discovery.getAccountInfo(account);
        //         discovery.stop();
        //         deferred.resolve(this._response(account));
        //     }
        // });
        // discovery.on('complete', () => {
        //     deferred.resolve(this._response(null));
        // });
        //
        // discovery.start();
        //
        // return await deferred.promise;
    }

    async getBitcoinHDNode(
      path: Array<number>,
      coinInfo?: ?CoinInfo
    ): Promise<trezor.HDNodeResponse> {
      const suffix: number = 0;
      const childPath: Array<number> = path.concat([suffix]);


      const resKey: MessageResponse<trezor.PublicKey> =
        await this.getPublicKey(path, 'Bitcoin');
      const childKey: MessageResponse<trezor.PublicKey>=
        await this.getPublicKey(childPath, 'Bitcoin');
      const publicKey: trezor.PublicKey =
        hdnodeUtils.xpubDerive(resKey, childKey, suffix);

      // const resKey: trezor.PublicKey = await this.session.getPublicKey(path, 'Bitcoin');
      // const childKey: trezor.PublicKey = await this.session.getPublicKey(childPath, 'Bitcoin');
      // const publicKey: trezor.PublicKey = hdnodeUtils.xpubDerive(resKey, childKey, suffix);

      const response: trezor.HDNodeResponse = {
        path,
        serializedPath: getSerializedPath(path),
        childNum: publicKey.node.child_num,
        xpub: coinInfo ? hdnodeUtils.convertBitcoinXpub(publicKey.xpub, coinInfo.network) : publicKey.xpub,
        chainCode: publicKey.node.chain_code,
        publicKey: publicKey.node.public_key,
        fingerprint: publicKey.node.fingerprint,
        depth: publicKey.node.depth,
      };

      // if requested path is a segwit
      // convert xpub to new format
      if (coinInfo) {
        const segwitNetwork = getSegwitNetwork(coinInfo);
        if (segwitNetwork && isSegwitPath(path)) {
          response.xpubSegwit = hdnodeUtils.convertBitcoinXpub(publicKey.xpub, segwitNetwork);
        }
      }
      return response;
    }

  async getPublicKey(
    address_n: Array<number>,
    coin_name: string,
    script_type?: ?string,
  ): Promise<trezor.PublicKey> {
    const response: MessageResponse<trezor.PublicKey> =
      await this.session.typedCall('GetPublicKey', 'PublicKey', {
      address_n,
      coin_name,
      script_type,
    });
    return response.message;
  }

    async _getHDNode(path: Array<number>, coinInfo: ?CoinInfo): Promise<trezor.HDNodeResponse> {
      // return this._getBitcoinHDNode(path, coinInfo);
      // if (!this.device.atLeast(['1.7.2', '2.0.10']))
      const isOld = true;
      if (isOld) {
        return await this.getBitcoinHDNode(path, coinInfo);
      }
      if (!coinInfo) {
        return await this.getBitcoinHDNode(path);
      }

      const suffix: number = 0;
      const childPath: Array<number> = path.concat([suffix]);
      let network: ?bitcoin.Network;
      if (isMultisigPath(path)) {
        network = coinInfo.network;
      } else if (isSegwitPath(path)) {
        network = getSegwitNetwork(coinInfo);
      } else if (isBech32Path(path)) {
        network = getBech32Network(coinInfo);
      }

      let scriptType: ?string = getScriptType(path);
      if (!network) {
        network = coinInfo.network;
        if (scriptType !== 'SPENDADDRESS') {
          scriptType = undefined;
        }
      }

      // const resKeyMessage: MessageResponse<trezor.PublicKey> =
      //   await this.getPublicKey(path, coinInfo.name, scriptType);
      // const resKey: trezor.PublicKey = resKeyMessage.message;
      //
      // const childKeyMessage: MessageResponse<trezor.PublicKey>=
      //   await this.getPublicKey(childPath, coinInfo.name, scriptType);
      // const childKey: trezor.PublicKey = childKeyMessage.message;
      //
      // const publicKey: trezor.PublicKey =
      //   hdnodeUtils.xpubDerive(resKey, childKey, suffix, network, coinInfo.network);

      const resKey: MessageResponse<trezor.PublicKey> =
        await this.getPublicKey(path, coinInfo.name, scriptType);
      const childKey: MessageResponse<trezor.PublicKey>=
        await this.getPublicKey(childPath, coinInfo.name, scriptType);
      const publicKey: trezor.PublicKey =
        hdnodeUtils.xpubDerive(resKey, childKey, suffix, network, coinInfo.network);

      const response: trezor.HDNodeResponse = {
        path,
        serializedPath: getSerializedPath(path),
        childNum: publicKey.node.child_num,
        // xpub: coinInfo ? hdnodeUtils.convertBitcoinXpub(publicKey.xpub, coinInfo.network) : publicKey.xpub,
        xpub: publicKey.xpub,
        chainCode: publicKey.node.chain_code,
        publicKey: publicKey.node.public_key,
        fingerprint: publicKey.node.fingerprint,
        depth: publicKey.node.depth,
      };

      // if requested path is a segwit
      // convert xpub to new format
      // if (coinInfo) {
      //   const segwitNetwork = getSegwitNetwork(coinInfo);
      //   if (segwitNetwork && isSegwitPath(path)) {
      //     response.xpubSegwit = hdnodeUtils.convertBitcoinXpub(publicKey.xpub, segwitNetwork);
      //   }
      // }

      if (network !== coinInfo.network) {
        response.xpubSegwit = response.xpub;
        response.xpub = hdnodeUtils.convertXpub(publicKey.xpub, network, coinInfo.network);
      }
      return response;
    }

    async _getAccountFromDiscovery(): Promise<Response> {
        const discovery: Discovery = this.discovery = new Discovery({
            // 替换
            // getHDNode: this.device.getCommands().getHDNode.bind(this.device.getCommands()),
            getHDNode: this._getHDNode,
            coinInfo: this.params.coinInfo,
            backend: this.backend,
        });

        const discoveryPromise: Deferred<void> = createDeferred();

        discovery.on('update', (accounts: Array<Account>) => {
            // this.postMessage(new UiMessage(UI.SELECT_ACCOUNT, {
            //     coinInfo: this.params.coinInfo,
            //     accounts: accounts.map(a => a.toMessage()),
            // }));
          console.log('discovery.on update')
        });

        discovery.on('complete', (accounts: Array<Account>) => {
            // this.postMessage(new UiMessage(UI.SELECT_ACCOUNT, {
            //     coinInfo: this.params.coinInfo,
            //     accounts: accounts.map(a => a.toMessage()),
            //     complete: true,
            // }));
          console.log('discovery.on complete');
          discoveryPromise.resolve();
        });

        try {
            discovery.start();
        } catch (error) {
            return {
                error,
            };
        }

        // set select account view
        // this view will be updated from discovery events
        // this.postMessage(new UiMessage(UI.SELECT_ACCOUNT, {
        //     coinInfo: this.params.coinInfo,
        //     accounts: [],
        //     start: true,
        // }));

        // wait for user action
        // const uiResp: UiPromiseResponse = await this.createUiPromise(UI.RECEIVE_ACCOUNT, this.device).promise;

        await discoveryPromise.promise;
        discovery.stop();

        // const resp: number = parseInt(uiResp.payload);
        console.log(`discovery accounts count: ${discovery.accounts.length}`)
        const account = discovery.accounts[0];

        return this._response(account);
    }

    _response(account: ?Account): Response {
        if (!account) {
            return {
                error: 'No account found',
            };
        }

        const nextAddress: string = account.getNextAddress();

        return {
            id: account.id,
            path: account.path,
            serializedPath: getSerializedPath(account.path),
            address: nextAddress,
            addressIndex: account.getNextAddressId(),
            addressPath: account.getAddressPath(nextAddress),
            addressSerializedPath: getSerializedPath(account.getAddressPath(nextAddress)),
            xpub: account.xpub,
            balance: account.getBalance(),
            confirmed: account.getConfirmedBalance(),
        };
    }

    dispose() {
        if (this.discovery) {
            const d = this.discovery;
            d.stop();
            d.removeAllListeners();
        }
    }
}
