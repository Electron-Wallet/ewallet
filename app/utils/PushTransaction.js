/* @flow */

import { validateParams } from './helpers/paramsValidator';
import { getCoinInfo } from 'utils/data/CoinInfo';
import { NO_COIN_INFO } from 'constants/errors';
import { create as createBlockbookBackend } from 'utils/backend';
import type { CoinInfo } from 'utils/types';

type Params = {
    tx: string,
    coinInfo: CoinInfo,
}

export type ComposeTransactionOptions = {
  tx: string,
  coin: string,
};

export default class PushTransaction {
    params: Params;

    constructor(options: ComposeTransactionOptions) {
        // validate incoming parameters
        validateParams(options, [
            { name: 'tx', type: 'string', obligatory: true },
            { name: 'coin', type: 'string', obligatory: true },
        ]);

        const coinInfo: ?CoinInfo = getCoinInfo(options.coin);
        if (!coinInfo) {
            throw NO_COIN_INFO;
        }

        if (coinInfo.type === 'bitcoin' && !/^[0-9A-Fa-f]*$/.test(options.tx)) {
            throw new Error('Invalid params: Transaction must be hexadecimal');
        }

        this.params = {
            tx: options.tx,
            coinInfo,
        };
    }

    async run(): Promise<{ txid: string }> {
      return await this.pushBlockbook();
    }

    async pushBlockbook(): Promise<{ txid: string }> {
        const { coinInfo } = this.params;
        const backend = await createBlockbookBackend(coinInfo);
        const txid: string = await backend.sendTransactionHex(this.params.tx);
        return {
            txid,
        };
    }
}
