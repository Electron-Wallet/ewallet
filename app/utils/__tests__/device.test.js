import * as dUtils from 'utils/device';

describe('device utils', () => {
    it('get status', () => {
        const deviceMock = [
            {
                connected: false,
            },
            {
                connected: true,
                available: false,
            },
            {
                connected: true,
                available: false,
                type: null,
            },
            {
                connected: true,
                available: true,
                type: 'acquired',
            },
            {
                connected: true,
                available: true,
                type: 'unacquired',
            },
            {
                connected: true,
                available: true,
                type: 'acquired',
                status: 'occupied',
            },
        ];

        deviceMock.forEach((device) => {
            expect(dUtils.getStatus(device)).toMatchSnapshot();
        });
    });

    it('isWebUSB', () => {
        const data = [
            { transport: { type: 'ParallelTransport', version: 'webusb' } },
            { transport: { type: null, version: 'aaaaaa' } },
            { transport: { type: 'ParallelTransport', version: 'webusb' } },
        ];

        data.forEach((item) => {
            expect(dUtils.isWebUSB(item.transport)).toMatchSnapshot();
        });
    });

    it('isDisabled', () => {
        const data = [
            { selectedDevice: { features: null }, devices: [1, 2, 3], transport: { version: 'webusb' } },
            { selectedDevice: { features: null }, devices: [], transport: { version: 'test' } },
        ];

        data.forEach((item) => {
            expect(dUtils.isDisabled(item.selectedDevice, item.devices, item.transport)).toMatchSnapshot();
        });
    });

    it('get version', () => {
        const deviceMock = [
            { },
            { features: {} },
            { features: { major_version: null } },
            { features: { major_version: 0 } },
            { features: { major_version: 1 } },
            { features: { major_version: 2 } },
        ];

        deviceMock.forEach((device) => {
            expect(dUtils.getVersion(device)).toMatchSnapshot();
        });
    });

    it('get status color', () => {
        const entry = [
            0,
            null,
            'sdsdsdsd',
            'used-in-other-window',
            'connected',
            'unacquired',
            'disconnected',
            'unavailable',
        ];

        entry.forEach((status) => {
            expect(dUtils.getStatusColor(status)).toMatchSnapshot();
        });
    });

    it('get status name', () => {
        const entry = [
            0,
            null,
            'sdsdsdsd',
            'used-in-other-window',
            'connected',
            'unacquired',
            'disconnected',
            'unavailable',
        ];

        entry.forEach((status) => {
            expect(dUtils.getStatusName(status)).toMatchSnapshot();
        });
    });
});
