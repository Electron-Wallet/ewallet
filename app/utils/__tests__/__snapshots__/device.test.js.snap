// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`device utils get status 1`] = `"disconnected"`;

exports[`device utils get status 2`] = `"unavailable"`;

exports[`device utils get status 3`] = `"unavailable"`;

exports[`device utils get status 4`] = `"connected"`;

exports[`device utils get status 5`] = `"unacquired"`;

exports[`device utils get status 6`] = `"used-in-other-window"`;

exports[`device utils get status color 1`] = `"#494949"`;

exports[`device utils get status color 2`] = `"#494949"`;

exports[`device utils get status color 3`] = `"#494949"`;

exports[`device utils get status color 4`] = `"#EB8A00"`;

exports[`device utils get status color 5`] = `"#01B757"`;

exports[`device utils get status color 6`] = `"#EB8A00"`;

exports[`device utils get status color 7`] = `"#ED1212"`;

exports[`device utils get status color 8`] = `"#ED1212"`;

exports[`device utils get status name 1`] = `"Status unknown"`;

exports[`device utils get status name 2`] = `"Status unknown"`;

exports[`device utils get status name 3`] = `"Status unknown"`;

exports[`device utils get status name 4`] = `"Used in other window"`;

exports[`device utils get status name 5`] = `"Connected"`;

exports[`device utils get status name 6`] = `"Used in other window"`;

exports[`device utils get status name 7`] = `"Disconnected"`;

exports[`device utils get status name 8`] = `"Unavailable"`;

exports[`device utils get version 1`] = `"1"`;

exports[`device utils get version 2`] = `"1"`;

exports[`device utils get version 3`] = `"1"`;

exports[`device utils get version 4`] = `"1"`;

exports[`device utils get version 5`] = `"1"`;

exports[`device utils get version 6`] = `"T"`;

exports[`device utils isDisabled 1`] = `false`;

exports[`device utils isDisabled 2`] = `true`;

exports[`device utils isWebUSB 1`] = `true`;

exports[`device utils isWebUSB 2`] = `false`;

exports[`device utils isWebUSB 3`] = `true`;
